modules/build/src/main/scala/scala/build/Inputs.scala

```diff
@@ -217,41 +222,50 @@ object Inputs {
     cwd: os.Path,
     directories: Directories,
     baseProjectName: String,
-    download: String => Either[String, Array[Byte]],
+    download: String => Either[String, List[UrlPath]],
     stdinOpt: => Option[Array[Byte]],
     acceptFds: Boolean
   ): Either[String, Inputs] = {
-    val validatedArgs = args.zipWithIndex.map {
+    val validatedArgs: Seq[Either[String, Seq[Compiled]]] = args.zipWithIndex.map {
       case (arg, idx) =>
         lazy val path      = os.Path(arg, cwd)
         lazy val dir       = path / os.up
         lazy val subPath   = path.subRelativeTo(dir)
         lazy val stdinOpt0 = stdinOpt
         val isStdin = (arg == "-" || arg == "-.scala" || arg == "_" || arg == "_.scala") &&
           stdinOpt0.nonEmpty
-        if (isStdin) Right(VirtualScalaFile(stdinOpt0.get, "<stdin>"))
+        if (isStdin) Right(Seq(VirtualScalaFile(stdinOpt0.get, "<stdin>")))
         else if ((arg == "-.sc" || arg == "_.sc") && stdinOpt0.nonEmpty)
-          Right(VirtualScript(stdinOpt0.get, "<stdin>", os.sub / "stdin.sc"))
+          Right(Seq(VirtualScript(stdinOpt0.get, "<stdin>", os.sub / "stdin.sc")))
         else if (arg.contains("://"))
-          download(arg).map { content =>
-            val wrapperPath = {
-              val u = new URI(arg) // FIXME Ignore parsing errors?
-              val it = Option(u.getScheme).iterator ++
-                Option(u.getAuthority).iterator ++
-                Option(u.getPath).iterator.flatMap(_.split('/').iterator)
-              os.sub / it.filter(_.nonEmpty).toVector
+          download(arg).map { files =>
```

Comment:
The Gist detection could be done around here. If a Gist URL is found, it can be changed to the Gist zip file (see the "Download ZIP" button on every Gist), and the zip file be downloaded here.

With the zip file, we can have a look at its content without unzipping it on disk, using the `java.util.zip.*` APIs (`ZipInputStream`, that can be passed a `ByteArrayInputStream` built from the `Array[Byte]`, or `ZipFile`, that can be passed a `File` if we change `download` to give us an `os.Path`)

This way, we can peek files from the Gist zip, and create `Virtual*` instances out of it, with a sensible `source` field (for `source`, we can try to build URLs like [this one](https://gist.github.com/alexarchambault/7b4ec20c4033690dd750ffd601e540ec#file-main-sc), pointing at the Gist individual files).